c10/xpu/XPUFunctions.cpp

```diff
@@ -0,0 +1,285 @@
+#include <c10/util/CallOnce.h>
+#include <c10/util/Exception.h>
+#include <c10/xpu/XPUFunctions.h>
+
+#ifndef _WIN32
+#include <sys/wait.h>
+#include <unistd.h>
+#endif
+#include <cmath>
+#include <deque>
+#include <mutex>
+#include <vector>
+
+namespace c10::xpu {
+namespace {
+
+/*
+ * Note [Device Management]
+ *
+ * Intel GPU device is based on sycl::device enumerated via SYCL runtime and
+ * device runtime status can be managed via a sycl device pool. The number of
+ * GPU devices is determined at run time.
+ *
+ * Currently, there is one SYCL device pool and the device pool is lazily
+ * created only once. The device management mechanism is thread local safe.
+ * The same default sycl context can shared for each sycl device.
+ *
+ * Device properties are initialized via the specific raw device.
+ */
+static c10::once_flag init_flag;
+static thread_local DeviceIndex curDeviceIndex = 0;
+
+struct DevicePool {
+  std::vector<std::unique_ptr<sycl::device>> devices;
+  std::unique_ptr<sycl::context> contexts;
+} gDevicePool;
+
+static void enumDevices(std::vector<std::unique_ptr<sycl::device>>& devices) {
+  auto platform_list = sycl::platform::get_platforms();
+  // Enumerated GPU devices from the specific platform.
+  for (const auto& platform : platform_list) {
+    if (platform.get_backend() != sycl::backend::ext_oneapi_level_zero) {
+      continue;
+    }
+    auto device_list = platform.get_devices();
+    for (const auto& device : device_list) {
+      if (device.is_gpu()) {
+        devices.push_back(std::make_unique<sycl::device>(device));
+      }
+    }
+  }
+}
+
+static inline int deviceCountImpl(
+    std::vector<std::unique_ptr<sycl::device>>& devices) {
+  enumDevices(devices);
+  return static_cast<int>(devices.size());
+}
+
+static inline void initGlobalDevicePoolState() {
+  // Get device count and record the all GPU devices.
+  auto device_count = deviceCountImpl(gDevicePool.devices);
+  if (device_count <= 0) {
+    TORCH_WARN("XPU device count is zero!");
+  }
+
+  // Here we use default context provided by Intel oneapi extension for each
+  // Intel GPU device.
+  gDevicePool.contexts = std::make_unique<sycl::context>(
+      gDevicePool.devices[0]->get_platform().ext_oneapi_get_default_context());
+}
+
+static inline void initDevicePoolCallOnce() {
+  c10::call_once(init_flag, initGlobalDevicePoolState);
+}
+
+static void initDeviceProperties(DeviceProp* device_prop, int device) {
+  using namespace sycl::info;
+  using namespace sycl::ext;
+  // Get raw sycl device associated with device index.
+  auto& raw_device = *gDevicePool.devices[device];
+
+  // clang-format off
+  // Initialize the device properties associated with the specific device.
+  device_prop->device_name = raw_device.get_info<device::name>();
+  device_prop->device_type = raw_device.get_info<device::device_type>();
+  device_prop->platform_name = raw_device.get_info<device::platform>().get_info<platform::name>();
```

Comment:
The style of this line is different from others. Does it mean we cannot get the platform name directly like `raw_device.get_info<device::platform_name>`?

Reply:
yes, platform name need to be fetched from sycl platform.